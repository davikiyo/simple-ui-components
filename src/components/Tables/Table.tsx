import { useEffect, useMemo, useState } from 'react'

import { styled, CSS } from 'styles'
import { sortObjects } from 'utils/utility'

import type { DataKeyType, TableData } from './models/table'
import TableBody, { TableField } from './TableBody'
import TableHead from './TableHead'
import TableRow from './TableRow'

export type SortKeyType = {
  key: string
  order: SORT_ORDER
}

export interface TableFieldType<T> extends TableField {
  key: DataKeyType<T>
  renderCell?: (data: TableData<T>) => JSX.Element | string
}

export interface TableProps<T> {
  /**
   * Overrides the style in the table container.
   */
  css?: CSS

  /**
   * Overrides the style in the table container.
   */
  tableCss?: CSS

  /**
   * Defines a class for table's container.
   */
  className?: string

  /**
   * The table data. Matches the key in the fields.
   * If no key is found in the `fields`, the value will be ignored.
   */
  data: TableData<T>[]

  /**
   * The property to be used as a key. (Default: `id`)
   *
   * **NOTE:** You can provide a key delimited by period (.) to get
   * the recursive item up to 5 recursion.
   *
   * @example
   * An example using `symbol` in `company`.
   * ```
   * const data = [
   *  {
   *    id: "C1",
   *    company: {
   *      name: "Example"
   *      symbol: "EXAM"
   *    }
   *  }
   * ]
   *  <Table
   *    dataKey="company.symbol"
   *   //...
   *  />
   * ```
   */
  dataKey?: DataKeyType<T>

  /**
   * The table field information. Matches the key in `data`.
   * If no `title` is provided, the `key` properties are displayed as titles.
   * The fields becomes sortable by providing the `sortable` flag.
   *
   * **NOTE:** You can provide a key delimited by period (.) to get
   * the nested item. (Up to depth of 5)
   *
   * @example
   * An example using `symbol` in `company`.
   * ```
   * const data = [
   *  {
   *    id: "C1",
   *    company: {
   *      name: "Example"
   *      symbol: "EXAM"
   *    }
   *  }
   * ]
   * const fields = [
   *  {
   *    key: "company.name",
   *    title: "Company Name"
   *  },
   *  {
   *    key: "company.symbol",
   *    title: "Symbol"
   *  }
   * ]
   *  <Table
   *    data={data}
   *    fields={fields}
   *    //...
   *  />
   * ```
   */
  fields: TableFieldType<T>[]

  /**
   * Defines the paddings for each cell in the table.
   */
  paddings?: number

  /**
   * Handles the event when clicked on sort buttons. The function passes the
   * current sort key as an argument.
   *
   * **NOTE:** If the function is omitted, the table itself will handle the sorting.
   * @param sortKey - A key of the column that user pressed.
   */
  onSortRequest?: (sortKey: SortKeyType) => void

  /**
   * Defines the height of each row in the table.
   */
  rowHeight?: number

  /**
   * Defines the width of the table.
   */
  width?: number | string

  /**
   * Defines the height of the table.
   */
  height?: number | string

  /**
   * Display the header in a sticky manner.
   */
  stickyHeader?: boolean

  /**
   * Display the header in a vertical manner.
   */
  verticalHeader?: boolean
}

const Container = styled('div', {
  fontFamily: '$main',
  display: 'inline-block',
  padding: 0,
  margin: 0,
  overflow: 'auto',
})

const TableContainer = styled('table', {
  border: 'none',
  borderSpacing: 0,
})

const TableHeadContainer = styled('thead', {
  margin: 0,
  backgroundColor: '#f0f0f0',
  zIndex: 10,
  variants: {
    stickyHeader: {
      true: {
        position: 'sticky',
        top: 0,
      },
    },
  },
})

export enum SORT_ORDER {
  ASC = 'ASC',
  DESC = 'DESC',
}

/** displays the given data in a styled table. */
export default function Table<T>({
  className,
  css,
  tableCss,
  data,
  dataKey = 'id',
  fields,
  rowHeight,
  height,
  width,
  onSortRequest,
  paddings = 16,
  stickyHeader = false,
  verticalHeader = false,
}: TableProps<T>) {
  const [sortKey, setSortKey] = useState<SortKeyType>({
    key: '',
    order: SORT_ORDER.ASC,
  })

  const tableData = useMemo(() => {
    if (!onSortRequest) {
      const sortedData = [...data]
      sortObjects(sortedData, sortKey.key, sortKey.order)
      return sortedData
    }

    return data
  }, [data, sortKey])

  useEffect(() => {
    if (sortKey.key && onSortRequest) onSortRequest(sortKey)
  }, [sortKey])

  const onSortHandler = (title: string) => {
    const key = fields.find((field) => field.key === title || field.title === title)?.key
    if (key) {
      const currentOrder =
        key === sortKey.key && sortKey.order === SORT_ORDER.ASC ? SORT_ORDER.DESC : SORT_ORDER.ASC

      setSortKey({ key, order: currentOrder })
    }
  }

  return (
    <Container
      className={className}
      css={{
        ...css,
        height,
        width,
      }}
      {...((height || width) && { tabIndex: 0 })}
    >
      <TableContainer css={{ width, height, ...tableCss }}>
        {!verticalHeader && (
          <TableHeadContainer stickyHeader={stickyHeader}>
            <TableRow height={rowHeight}>
              {fields.map(({ key, title, sortable }) => (
                <TableHead
                  // Assign key & title when there are duplicate keys
                  key={`${key}_${title || ''}`}
                  sortable={sortable}
                  paddings={paddings}
                  stickyHeader={stickyHeader}
                  {...(sortable && {
                    onSortClick: onSortHandler,
                  })}
                  {...(key === sortKey.key && {
                    sortOrder: sortKey.order,
                  })}
                >
                  {title || key}
                </TableHead>
              ))}
            </TableRow>
          </TableHeadContainer>
        )}
        <TableBody
          data={tableData}
          dataKey={dataKey}
          fields={fields}
          rowHeight={rowHeight}
          paddings={paddings}
          verticalHeader={verticalHeader}
          stickyHeader={stickyHeader}
          sortKey={sortKey}
          onSortClick={onSortHandler}
        />
      </TableContainer>
    </Container>
  )
}
