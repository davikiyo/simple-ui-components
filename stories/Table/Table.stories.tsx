import { Meta, StoryObj } from '@storybook/react'

import { Table, TableProps, Card, Pagination, IconButton, SelectedRowKey } from 'components'
import { useState } from 'react'
import { Paginator } from 'components'

const meta: Meta<typeof Table> = {
  component: Table,
  args: {
    onSortRequest: undefined,
  },
}

export default meta

type TestDataType = {
  id: number
  title: string
  author: string
  price: number
  publisher: {
    name: string
    publisherId: string
  }
}

const sampleData: TestDataType[] = [
  {
    id: 1,
    title: 'Test 1',
    author: 'Author 1',
    price: 5.0,
    publisher: {
      name: 'Publisher 1',
      publisherId: 'p1',
    },
  },
  {
    id: 2,
    title: 'Test 2',
    author: 'Author 2',
    price: 7.99,
    publisher: {
      name: 'Publisher 1',
      publisherId: 'p1',
    },
  },
  {
    id: 3,
    title: 'Test 3',
    author: 'Author 1',
    price: 11.99,
    publisher: {
      name: 'Publisher 2',
      publisherId: 'p2',
    },
  },
  {
    id: 4,
    title: 'Test 4',
    author: 'Author 3',
    price: 9.99,
    publisher: {
      name: 'Publisher 2',
      publisherId: 'p2',
    },
  },
]

type Story = StoryObj<typeof Table<TestDataType>>

export const NonSortable: Story = {
  args: {
    data: sampleData,
    fields: [
      {
        key: 'title',
        title: 'Title',
      },
      {
        key: 'author',
        title: 'Author',
      },
      {
        key: 'price',
        title: 'Price',
      },
      {
        key: 'publisher.name',
        title: 'Publisher',
      },
    ],
  },
}

export const Sortable: Story = {
  args: {
    ...NonSortable.args,
    fields: [
      {
        key: 'title',
        title: 'Title',
        sortable: true,
      },
      {
        key: 'author',
        title: 'Author',
        sortable: true,
      },
      {
        key: 'price',
        title: 'Price',
        sortable: true,
      },
      {
        key: 'publisher.name',
        title: 'Publisher',
        sortable: true,
      },
    ],
  },
}

export const Large: Story = {
  args: {
    ...NonSortable.args,
    rowHeight: 70,
  },
}

export const Small: Story = {
  args: {
    ...NonSortable.args,
    rowHeight: 20,
    paddings: 8,
  },
}

export const FixedHeightAndWidth: Story = {
  args: {
    ...NonSortable.args,
    height: 200,
    width: 250,
  },
}

export const StickyHeader: Story = {
  args: {
    ...NonSortable.args,
    data: sampleData.concat([
      {
        id: 5,
        title: 'Test 5',
        author: 'Author 3',
        price: 10.99,
        publisher: {
          name: 'Publisher 3',
          publisherId: 'p3',
        },
      },
      {
        id: 6,
        title: 'Test 6',
        author: 'Author 3',
        price: 10.99,
        publisher: {
          name: 'Publisher 3',
          publisherId: 'p3',
        },
      },
      {
        id: 7,
        title: 'Test 7',
        author: 'Author 3',
        price: 10.99,
        publisher: {
          name: 'Publisher 3',
          publisherId: 'p3',
        },
      },
    ]),
    stickyHeader: true,
    height: 250,
  },
}

export const VerticalHeader: Story = {
  args: {
    ...Sortable.args,
    verticalHeader: true,
  },
}

export const StickyVerticalHeader: Story = {
  args: {
    ...VerticalHeader.args,
    width: 300,
    stickyHeader: true,
  },
}

export const CustomCells: Story = {
  args: {
    data: sampleData,
    fields: [
      {
        key: 'title',
        title: 'Title',
      },
      {
        key: 'author',
        title: 'Author',
      },
      {
        key: 'price',
        title: 'Price',
      },
      {
        key: 'publisher.name',
        title: 'Publisher',
      },
      {
        key: 'id',
        title: 'Action',
        renderCell: (data) => {
          return <button>Action {data.title}</button>
        },
      },
      {
        key: 'id',
        title: 'Delete',
        renderCell: (data) => {
          return <button>Delete {data.title}</button>
        },
      },
    ],
  },
}

export const SortableWithHandler: Story = {
  args: {
    ...NonSortable.args,
    fields: [
      {
        key: 'title',
        title: 'Title',
        sortable: true,
      },
      {
        key: 'author',
        title: 'Author',
        sortable: true,
      },
      {
        key: 'price',
        title: 'Price',
        sortable: true,
      },
      {
        key: 'publisher.name',
        title: 'Publisher',
        sortable: true,
      },
    ],
    onSortRequest: (sortKey) => {
      console.log(sortKey)
    },
  },
}

type IntegratedDataType = {
  id: number
  name: string
  code: string
  price: string
  pe: number
  isFavorite: boolean
}

const data = Array.from({ length: 100 }, (_, i) => ({
  id: i + 1,
  name: `Stock ${i + 1}`,
  code: `STOCK${i + 1}`,
  price: '999.99',
  pe: (i + 1) / 10,
  isFavorite: i === 0 ? true : false,
}))

const MAX_PAGE_SIZE = 20

const IntegratedComponent = (props: TableProps<IntegratedDataType>) => {
  const [pageNumber, setPageNumber] = useState(1)
  const [tableData, setTableData] = useState(data.slice(0, MAX_PAGE_SIZE))

  return (
    <div
      style={{
        display: 'flex',
        alignItems: 'center',
      }}
    >
      <Card
        css={{
          marginLeft: 'auto',
          marginRight: 'auto',
          paddingLeft: 0,
          paddingRight: 0,
          width: '100%',
          height: '100%',
          maxWidth: '768px',
        }}
        border={false}
        shadow
      >
        <Table stickyHeader css={{ maxHeight: '387px' }} width="100%" {...props} data={tableData} />
        <div
          style={{
            display: 'flex',
            marginTop: '16px',
            marginBottom: '16px',
            justifyContent: 'center',
          }}
        >
          <Pagination
            pages={Math.ceil(data.length / MAX_PAGE_SIZE)}
            currentPage={pageNumber}
            onNextClick={() => {
              const newPageNum = pageNumber + 1
              setTableData(Paginator.paginate(data, MAX_PAGE_SIZE, newPageNum))
              setPageNumber(newPageNum)
            }}
            onPrevClick={() => {
              const newPageNum = pageNumber - 1
              setTableData(Paginator.paginate(data, MAX_PAGE_SIZE, newPageNum))
              setPageNumber(newPageNum)
            }}
            onPageClick={(page) => {
              setTableData(Paginator.paginate(data, MAX_PAGE_SIZE, page))
              setPageNumber(page)
            }}
          />
        </div>
      </Card>
    </div>
  )
}
type IntegratedStory = StoryObj<typeof Table<IntegratedDataType>>
export const Integrated: IntegratedStory = {
  args: {
    fields: [
      {
        key: 'name',
        title: 'Name',
        sortable: true,
      },
      {
        key: 'code',
        title: 'Code',
        sortable: true,
      },
      {
        key: 'price',
        title: 'Price',
        sortable: true,
      },
      {
        key: 'pe',
        title: 'P/E',
        sortable: true,
      },
      {
        key: 'isFavorite',
        title: 'Favorite',
        sortable: true,
        renderCell: ({ isFavorite }) =>
          isFavorite ? (
            <IconButton icon="star-filled" color="#FFC107" height={24} width={24} />
          ) : (
            <IconButton icon="star-outlined" color="#FFC107" height={24} width={24} />
          ),
      },
    ],
  },
  render: (args) => <IntegratedComponent {...args} />,
}

export const SelectableRows: Story = {
  args: {
    ...NonSortable.args,
    fields: [
      {
        key: 'title',
        title: 'Title',
      },
      {
        key: 'author',
        title: 'Author',
      },
      {
        key: 'price',
        title: 'Price',
      },
      {
        key: 'publisher.publisherId',
        title: 'Publisher',
        renderCell: (item) => (
          <div style={{ textAlign: 'left' }}>
            <div>Publisher ID: {item.publisher.publisherId}</div>
            <div>Publisher Name: {item.publisher.name}</div>
          </div>
        ),
      },
      {
        key: 'id',
        title: 'Action',
        renderCell: (data) => {
          return (
            <button
              onClick={() => {
                console.log(data.title)
              }}
            >
              Action {data.title}
            </button>
          )
        },
      },
    ],
    hoverable: true,
  },
  render: (args) => {
    const [selectedRows, setSelectedRows] = useState<SelectedRowKey[]>([])
    const onRowClickHandler = (key: SelectedRowKey) => {
      if (selectedRows.some((k) => k === key)) {
        setSelectedRows(selectedRows.filter((k) => k !== key))
      } else {
        setSelectedRows([...selectedRows, key])
      }
    }

    return <Table {...args} selectedRows={selectedRows} onRowClick={onRowClickHandler} />
  },
}

export const CollapsibleTable: Story = {
  args: {
    ...NonSortable.args,
    fields: [
      {
        key: 'title',
        title: 'Title',
      },
      {
        key: 'author',
        title: 'Author',
      },
      {
        key: 'price',
        title: 'Price',
      },
    ],
    hoverable: true,
    selectedRows: [1],
    renderRowDetail: (item) => (
      <>
        <div>
          <span>Publisher&apos;s ID: </span>
          <span>{item.publisher.publisherId}</span>
        </div>
        <div>
          <span>Publisher&apos;s Name: </span>
          <span>{item.publisher.name}</span>
        </div>
      </>
    ),
  },
  render: (args) => {
    const [selectedRows, setSelectedRows] = useState<SelectedRowKey[]>([])
    const onRowClickHandler = (key: SelectedRowKey) => {
      if (selectedRows.some((k) => k === key)) {
        setSelectedRows(selectedRows.filter((k) => k !== key))
      } else {
        setSelectedRows([key])
      }
    }

    return <Table {...args} selectedRows={selectedRows} onRowClick={onRowClickHandler} />
  },
}
