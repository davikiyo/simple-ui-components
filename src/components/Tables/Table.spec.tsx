import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import Table, { TableFieldType } from './Table'

interface TestDataType {
  id: number
  title: string
  author: string
  price: number
  publisher: {
    name: string
    id: string
  }
  coAuthors?: string[]
}

describe('Table component', () => {
  const onSortHandlerMock = jest.fn()
  let data: TestDataType[]
  let fields: TableFieldType<TestDataType>[]
  let vertical = false

  beforeEach(() => {
    data = [
      {
        id: 1,
        title: 'Title1',
        author: 'Author1',
        price: 10,
        publisher: {
          id: 'p1',
          name: 'Publisher1',
        },
      },
      {
        id: 2,
        title: 'Title2',
        author: 'Author2',
        price: 9,
        publisher: {
          id: 'p2',
          name: 'Publisher2',
        },
        coAuthors: ['Author3', 'Author4'],
      },
    ]
    fields = [
      {
        key: 'title',
        title: 'Title',
      },
      {
        key: 'author',
        title: 'Author',
      },
      {
        key: 'price',
        title: 'Price',
      },
      {
        key: 'publisher.name',
        title: 'Publisher',
      },
      {
        key: 'coAuthors',
        title: 'Co-Authors',
      },
    ]
    vertical = false
    onSortHandlerMock.mockClear()
  })

  const initialize = () =>
    render(
      <Table
        data={data}
        fields={fields}
        onSortRequest={onSortHandlerMock}
        verticalHeader={vertical}
      />
    )

  it('should render a table', () => {
    initialize()

    expect(screen.getByRole('table')).toBeInTheDocument()
  })

  it('should render a table with correct header', () => {
    initialize()

    fields.forEach((field) => {
      const title = field.title ? field.title : field.key
      expect(screen.getByText(title)).toBeInTheDocument()
    })
  })

  it('should render a table with correct data', () => {
    initialize()

    data.forEach((item) => {
      expect(screen.getByText(item.title)).toBeInTheDocument()
      expect(screen.getByText(item.author)).toBeInTheDocument()
      expect(screen.getByText(item.price)).toBeInTheDocument()
      expect(screen.getByText(item.publisher.name)).toBeInTheDocument()
    })
  })

  it('should render a sortable table head and trigger an event when clicked', async () => {
    fields[0].sortable = true
    initialize()

    await userEvent.click(screen.getByRole('button', { name: /Title/ }))
    expect(onSortHandlerMock).toHaveBeenCalled()
  })

  it('should sort the field when a handler is not provided', async () => {
    fields[0].sortable = true
    const { asFragment } = render(<Table data={data} fields={fields} />)

    await userEvent.click(screen.getByRole('button', { name: /Title/ }))
    await userEvent.click(screen.getByRole('button', { name: /Title/ }))
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render a custom cell', async () => {
    fields[0].renderCell = () => 'TEST'
    initialize()

    expect(screen.getAllByText('TEST')).toHaveLength(data.length)
  })

  it('should render a table with vertical headers', () => {
    vertical = true
    const { asFragment } = initialize()

    expect(screen.getByRole('table')).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render a sortable field in the table with vertical headers', async () => {
    vertical = true
    fields[0].sortable = true
    initialize()

    await userEvent.click(screen.getByRole('button', { name: /Title/ }))
    expect(onSortHandlerMock).toHaveBeenCalled()
  })

  it('should render a custom cell in the table with vertical headers', async () => {
    fields[0].renderCell = (item) => <button>{item.title}</button>
    vertical = true
    initialize()

    expect(screen.getByRole('button', { name: data[0].title })).toBeInTheDocument()
  })

  it('should render the key as a column header when there is no title', () => {
    fields[0].title = undefined
    vertical = true
    initialize()
    expect(screen.getByRole('columnheader', { name: fields[0].key })).toBeInTheDocument()
  })
})
