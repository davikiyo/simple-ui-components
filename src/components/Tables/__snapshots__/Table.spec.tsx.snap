// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component should render a table with vertical headers 1`] = `
<DocumentFragment>
  <div
    class="c-hMuChS c-hMuChS-iPJLV-css"
  >
    <table
      class="c-fQrKGu c-fQrKGu-iPJLV-css"
    >
      <tbody
        class="c-kXHQtC"
      >
        <tr
          class="c-PJLV c-PJLV-iPJLV-css"
        >
          <th
            class="c-dikTaN c-dikTaN-fJIFXR-vertical-true c-dikTaN-ikYLxwh-css"
          >
            Title
          </th>
          <td
            class="c-fppYJE c-fppYJE-ikYLxwh-css"
          >
            Title1
          </td>
          <td
            class="c-fppYJE c-fppYJE-ikYLxwh-css"
          >
            Title2
          </td>
        </tr>
        <tr
          class="c-PJLV c-PJLV-iPJLV-css"
        >
          <th
            class="c-dikTaN c-dikTaN-fJIFXR-vertical-true c-dikTaN-ikYLxwh-css"
          >
            Author
          </th>
          <td
            class="c-fppYJE c-fppYJE-ikYLxwh-css"
          >
            Author1
          </td>
          <td
            class="c-fppYJE c-fppYJE-ikYLxwh-css"
          >
            Author2
          </td>
        </tr>
        <tr
          class="c-PJLV c-PJLV-iPJLV-css"
        >
          <th
            class="c-dikTaN c-dikTaN-fJIFXR-vertical-true c-dikTaN-ikYLxwh-css"
          >
            Price
          </th>
          <td
            class="c-fppYJE c-fppYJE-ikYLxwh-css"
          >
            10
          </td>
          <td
            class="c-fppYJE c-fppYJE-ikYLxwh-css"
          >
            9
          </td>
        </tr>
        <tr
          class="c-PJLV c-PJLV-iPJLV-css"
        >
          <th
            class="c-dikTaN c-dikTaN-fJIFXR-vertical-true c-dikTaN-ikYLxwh-css"
          >
            Publisher
          </th>
          <td
            class="c-fppYJE c-fppYJE-ikYLxwh-css"
          >
            Publisher1
          </td>
          <td
            class="c-fppYJE c-fppYJE-ikYLxwh-css"
          >
            Publisher2
          </td>
        </tr>
        <tr
          class="c-PJLV c-PJLV-iPJLV-css"
        >
          <th
            class="c-dikTaN c-dikTaN-fJIFXR-vertical-true c-dikTaN-ikYLxwh-css"
          >
            Co-Authors
          </th>
          <td
            class="c-fppYJE c-fppYJE-ikYLxwh-css"
          />
          <td
            class="c-fppYJE c-fppYJE-ikYLxwh-css"
          >
            ["Author3","Author4"]
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`Table component should sort the field when a handler is not provided 1`] = `
<DocumentFragment>
  <div
    class="c-hMuChS c-hMuChS-iPJLV-css"
  >
    <table
      class="c-fQrKGu c-fQrKGu-iPJLV-css"
    >
      <thead
        class="c-eMJcLW"
      >
        <tr
          class="c-PJLV c-PJLV-iPJLV-css"
        >
          <th
            class="c-dikTaN c-dikTaN-ikYLxwh-css"
          >
            <button
              class="c-bkVFPF"
            >
              Title
              <span
                class="c-yVZdU"
              >
                <span
                  class="c-khZXrc c-khZXrc-ikDfHgf-css"
                >
                  <svg
                    aria-label="sorting-arrow-up"
                    fill="currentColor"
                    height="100%"
                    role="img"
                    width="100%"
                  >
                    <use
                      href="#simple-ui-icon-sorting-arrow-up"
                    />
                  </svg>
                </span>
                <span
                  class="c-khZXrc c-khZXrc-iiJqewt-css"
                >
                  <svg
                    aria-label="sorting-arrow-up"
                    fill="currentColor"
                    height="100%"
                    role="img"
                    width="100%"
                  >
                    <use
                      href="#simple-ui-icon-sorting-arrow-up"
                    />
                  </svg>
                </span>
              </span>
            </button>
          </th>
          <th
            class="c-dikTaN c-dikTaN-ikYLxwh-css"
          >
            Author
          </th>
          <th
            class="c-dikTaN c-dikTaN-ikYLxwh-css"
          >
            Price
          </th>
          <th
            class="c-dikTaN c-dikTaN-ikYLxwh-css"
          >
            Publisher
          </th>
          <th
            class="c-dikTaN c-dikTaN-ikYLxwh-css"
          >
            Co-Authors
          </th>
        </tr>
      </thead>
      <tbody
        class="c-kXHQtC"
      >
        <tr
          class="c-PJLV c-PJLV-iPJLV-css"
        >
          <td
            class="c-fppYJE c-fppYJE-ikYLxwh-css"
          >
            Title2
          </td>
          <td
            class="c-fppYJE c-fppYJE-ikYLxwh-css"
          >
            Author2
          </td>
          <td
            class="c-fppYJE c-fppYJE-ikYLxwh-css"
          >
            9
          </td>
          <td
            class="c-fppYJE c-fppYJE-ikYLxwh-css"
          >
            Publisher2
          </td>
          <td
            class="c-fppYJE c-fppYJE-ikYLxwh-css"
          >
            ["Author3","Author4"]
          </td>
        </tr>
        <tr
          class="c-PJLV c-PJLV-iPJLV-css"
        >
          <td
            class="c-fppYJE c-fppYJE-ikYLxwh-css"
          >
            Title1
          </td>
          <td
            class="c-fppYJE c-fppYJE-ikYLxwh-css"
          >
            Author1
          </td>
          <td
            class="c-fppYJE c-fppYJE-ikYLxwh-css"
          >
            10
          </td>
          <td
            class="c-fppYJE c-fppYJE-ikYLxwh-css"
          >
            Publisher1
          </td>
          <td
            class="c-fppYJE c-fppYJE-ikYLxwh-css"
          />
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
